x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java

```diff
@@ -326,16 +332,50 @@ protected void persistState(TransformState state, ActionListener<Void> listener)
                         "Failure updating stats of transform, unexpected version conflict of internal state, resetting to recover: "
                             + statsExc.getMessage()
                     );
-                    assert false : "[" + getJobId() + "] updating stats of transform failed, unexpected version conflict of internal state";
+
+                    if (handleStatePersistenceFailure(statsExc) == false) {
+                        // get the current seqNo and primary term, however ignore the stored state
+                        transformsConfigManager.getTransformStoredDoc(
+                            transformConfig.getId(),
+                            false,
+                            ActionListener.wrap(storedDocAndSeqNoPrimaryTerm -> {
+                                updateSeqNoPrimaryTermAndIndex(seqNoPrimaryTermAndIndex, storedDocAndSeqNoPrimaryTerm.v2());
```

Comment:
Why is this necessary if the previous storage succeeded?

We set the values when the previous call that caused THIS call to fail.

Reply:
The storage failed and this gets the current seq_no/primary_term needed to hopefully succeeds on the next state persistence. Otherwise state persistence endlessly tries to update the doc with the same outdated seq_no forever.

In case of 2 concurrent state persistence calls - which is just a theory and actually should not happen anymore - it won't have an effect, because in order to update the reference it only updates if the old value matches. If this happens it gets logged as a warning. Even in that case state persistence should recover.